nvbench/user_axis_space.cuh

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Copyright 2022 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/iteration_space_base.cuh>
+
+namespace nvbench
+{
+
+/*!
+ * Provides user defined iteration over multiple axes
+ *
+ * Consider two axi with the following values:
+ * { 0, 1, 2, 3, 4, 5 }
+ * { 0, 1, 2, 3, 4, 5 }
+ *
+ * If we wanted to provide an axis space that skipped every third value
+ * We would implement it like this:
+ *
+ * struct every_third final : nvbench::user_axis_space
+ * {
+ *   every_third(std::vector<std::size_t> input_indices,
+ *               std::vector<std::size_t> output_indices)
+ *       : nvbench::user_axis_space(std::move(input_indices),
+ *                                  std::move(output_indices))
+ *   {}
+ *
+ *   nvbench::detail::axis_space_iterator do_get_iterator(axes_info info) const
```

Comment:
Having things in the `detail` namespace appear in something called `user_axis_space` seems incongruous.

Reply:
I have updated the implementation around here, so it would be good to review again. If we are happy with `axis_space_iterator` I will promote it to `nvbench::`